### PR TITLE
Fix/issue90

### LINKED ISSUE #90: generate_from_csv: adjust min_standing_time
The min_standing_time is a fixed value now. I thought about adding a feature that calculates the min_standing_time from the vehicle battery capacity and the charging power. We could then assume that there needs to be enough time to load the vehicle at least e.g. 80% of it's soc. This might just work for a constant loading curve though. @jakobgemassmer ,what do you think?  

this is relevant for generate_from_csv.py and the bus_tool

### PR DESCRIPTION
**Pull request by Piranias**
Fix #90

**Changes proposed in this pull request**:
- add variable "recharge fraction" to calculate min_standing_time at depot

The following steps were realized, as well (required):
- [x] Correct linting with `fake8 path_to_script`
- [x] Check if tests pass locally (`pytest ./src/tests.py`)

Also the following steps were realized (if applies):
- [x] Write docstrings to your code
- [ ] For new functionalities: Explain in readthedocs
- [ ] Write test(s) for your new patch of code

*Please mark above checkboxes as following:*
- [ ] Open
- [x] Done

### CODE REVIEW

File: generate_from_csv.py [hunk 1]
```diff
@@ -359,71 +361,64 @@ def csv_to_dict(csv_path):
...
+    for rot in rotations:
+        # find vehicles that have completed rotation and stood for a minimum staning time
+        # mark those vehicle as idle
+        for r in rotations_in_progress:
+            # calculate min_standing_time deps
+            capacity = vehicle_types[r["vehicle_type"]]["capacity"]
+            cs_power = max([v[1] for v in vehicle_types[r["vehicle_type"]]['charging_curve']])
+            min_standing_time = (capacity / cs_power) * recharge_fraction
+
+            if datetime.datetime.strptime(rot["departure_time"], '%Y-%m-%d %H:%M:%S') > \
```
Comment (stefansc1) [inline generate_from_csv.py:396]: 1) repetition of datetime format -> make it a) variable (set in config?) or b) static (define format at beginning of file)
2) if-clause far too long to be readable. parse datetimes beforehand, give them nice variable names, also convert min_standing_time to timedelta:
```
departure_time = datetime.datetime.strptime(rot["departure_time"], DATETIME_FORMAT)
arrival_time = datetime.datetime.strptime(r["arrival_time"], DATETIME_FORMAT)
min_standing_time = datetime.timedelta(hours=min_standing_time)  # extract from loop

if departure_time - arrival_time > min_standing_time:
  ...
```

File: generate_from_csv.py [hunk 2]
```diff
@@ -359,71 +361,64 @@ def csv_to_dict(csv_path):
...
-                del arrivals[arrival_rotation]
-                input[rotation]["vehicle_id"] = a_bus_number
+    rotations_in_progress = []
+    idle_vehicles = []
+    vehicle_type_counts = {vehicle_type: 0 for vehicle_type in vehicle_types.keys()}
+
+    rotations = sorted(input, key=lambda d: d.get('departure_time'))
+
+    for rot in rotations:
+        # find vehicles that have completed rotation and stood for a minimum staning time
```
Comment (stefansc1) [inline generate_from_csv.py:388]: typo stan**d**ing time

File: generate_from_csv.py [hunk 3]
```diff
@@ -359,71 +361,64 @@ def csv_to_dict(csv_path):
...
+
+        # find idle vehicle for rotation if exists
+        # else generate new vehicle id
+        vt = rot["vehicle_type"]
+        id = next((id for id in idle_vehicles if vt in id), None)
+        if id is None:
+            vehicle_type_counts[vt] += 1
+            id = f"{vt}_{vehicle_type_counts[vt]}"
+        else:
+            idle_vehicles.remove(id)
```
Comment (stefansc1) [inline generate_from_csv.py:412]: more pythonic way:
```
try:
  id = next(id for id in idle_vehicles if vt in id)
  vehicle_type_counts[vt] += 1
  id = f"{vt}_{vehicle_type_counts[vt]}"
except StopIteration:
  idle_vehicles.remove(id)
```

**Review by stefansc1 — CHANGES_REQUESTED** [review]

File: examples/generate_from_csv.cfg [hunk 4]
```diff
@@ -23,11 +23,10 @@ v2g = false
 gc_power = 1500
 # set minimal power at charging station in kW
 cs_power_min = 0
-# set minimum standing time at depot in hours. If None, no min_standing_time is assigned
-min_standing_time = None
-# option to export csv after assigning vehicle_id ( file name in example directory, e.g. "vehicle_id_file.csv")
+# option to export csv after assigning vehicle_id (file name in example directory, e.g. "vehicle_id_file.csv") (default: None)
 export_vehicle_id_csv = None
-
+# minimum fraction of capacity for recharge when leaving the charging station. Helps to calculate the minimum min_standing_time. Only used if vehicle_id is not in input_file. [0,1] (default: 1)
```
Comment (jakobgemassmer) [inline examples/generate_from_csv.cfg:28]: ALternative to be more precise
```suggestion
# minimum fraction of vehicle battery capacity for recharge when leaving a charging station. Needed to calculate the minimum standing time at a charging station. Only used if vehicle_id is not in input_file. [0,1] (default: 1)
```

File: generate_from_csv.py [hunk 5]
```diff
@@ -359,71 +361,64 @@ def csv_to_dict(csv_path):
...
+    idle_vehicles = []
+    vehicle_type_counts = {vehicle_type: 0 for vehicle_type in vehicle_types.keys()}
+
+    rotations = sorted(input, key=lambda d: d.get('departure_time'))
+
+    for rot in rotations:
+        # find vehicles that have completed rotation and stood for a minimum staning time
+        # mark those vehicle as idle
+        for r in rotations_in_progress:
+            # calculate min_standing_time deps
```
Comment (jakobgemassmer) [inline generate_from_csv.py:391]: what is "deps"? This should work in general, not only for busses.
```suggestion
            # calculate min_standing_time at a charging station for each vehicle
```

File: generate_from_csv.py [hunk 6]
```diff
@@ -458,6 +452,9 @@ def assign_vehicle_id(input, min_standing_time, export=None):
     parser.add_argument('--cs-power-min', type=float, default=0, help='set minimal power at '
                                                                       'charging station in kW')
     parser.add_argument('--seed', default=None, type=int, help='set random seed')
+    parser.add_argument('--recharge-fraction', type=float, default=1,
+                        help='Minimum fraction of capacity for recharge when leaving the charging '
```
Comment (jakobgemassmer) [inline generate_from_csv.py:456]:
```suggestion
                        help='Minimum fraction of vehicle battery capacity for recharge when leaving the charging '
```

**Review by jakobgemassmer — COMMENTED** [review]
Approval for general approach. Only minor adjustments to be more precise.

**Review by stefansc1 — APPROVED** [review]

**Pull request merged**